akka-persistence-testkit/src/main/scala/akka/persistence/testkit/query/scaladsl/PersistenceTestKitReadJournal.scala

```diff
@@ -24,23 +26,51 @@ import akka.util.unused
 import com.typesafe.config.Config
 import org.slf4j.LoggerFactory
 import akka.persistence.Persistence
+import akka.persistence.PersistentRepr
+import akka.persistence.query.TimestampOffset
 import akka.persistence.query.typed
+import akka.persistence.query.typed.scaladsl.CurrentEventsByPersistenceIdTypedQuery
 import akka.persistence.query.typed.scaladsl.CurrentEventsBySliceQuery
+import akka.persistence.query.typed.scaladsl.EventsByPersistenceIdTypedQuery
 import akka.persistence.typed.PersistenceId
 
+import java.time.Instant
+import java.time.temporal.ChronoUnit
 import scala.collection.immutable
 
 object PersistenceTestKitReadJournal {
   val Identifier = "akka.persistence.testkit.query"
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi
+  private[akka] def tagsFor(payload: Any): Set[String] = payload match {
+    case Tagged(_, tags) => tags
+    case _               => Set.empty
+  }
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi
+  private[akka] def timestampOffsetFor(pr: PersistentRepr) = {
+    val timestamp = Instant.ofEpochMilli(pr.timestamp) // Note: we don't really have microsecond granularity here
```

Comment:
A good thing is that we always use an increasing and unique timestamp in the testkit, see `akka.persistence.testkit.internal.CurrentTime`